torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -176,20 +176,31 @@ void TensorPipeAgent::pipeRead(
     std::function<void(const tensorpipe::Error&, Message&&)> fn) {
   pipe->readDescriptor([fn{std::move(fn)}, pipe](
                            const tensorpipe::Error& error,
-                           tensorpipe::Message&& tpMessage) mutable {
+                           tensorpipe::Message tpMessage) mutable {
     if (error) {
       fn(error, Message());
       return;
     }
 
-    // Allocate memory and fill in pointers
-    Message rpcMessage = tensorpipeAllocateMessage(tpMessage);
+    TensorpipeReadBuffers tpBuffers = tensorpipeAllocate(tpMessage);
 
     pipe->read(
         std::move(tpMessage),
-        [fn{std::move(fn)}, rpcMessage{std::move(rpcMessage)}](
+        [tpBuffers{
+             std::make_shared<TensorpipeReadBuffers>(std::move(tpBuffers))},
+         fn{std::move(fn)}](
             const tensorpipe::Error& error,
-            tensorpipe::Message&& /* unused */) mutable {
+            tensorpipe::Message tpMessage) mutable {
+          if (error) {
```

Comment:
IIUC, this is the shortcut to avoid deserialization when there is an error?

Reply:
Yes, but this was actually almost unintentional. The real benefit will be had when the deserialization will be deferred to the thread pool.